jscomp/others/js_json.mli

```diff
@@ -30,7 +30,15 @@
 
 (** ## Types *)
 
-type t
+type t =
+  | False [@as false]
+  | True [@as true]
+  | Null [@as null]
+  | String of string
+  | Number of float
+  | Object of t Js.Dict.t
+  | Array of t array
+[@@unboxed]
```

Comment:
Shouldn't it be kept as an opaque type?

Reply:
Why?

Reply:
I think it's better to define a type as an opaque type, hiding its internal structure, so that values can only be manipulated using the functions provided by the module's interface, like the previous Js.Json.t.

Reply:
What's the advantage of hiding the fact that json is json?
Do you want to not rely on the fact that json is a standard? Perhaps there's some abstraction that's useful in some way? Not sure I get it. But curious about what's the thought.

Reply:
I'm also curious about why. I only see upsides, but maybe I'm missing something @mununki .

Reply:
I think it's better for a module's interface not to expose what doesn't need to be exposed. Actually, I don't really see the advantages clearly. On the other hand, what are the advantages of exposing Json.t this time?

<img width="828" alt="image" src="https://user-images.githubusercontent.com/36963424/235022649-a8f0e1be-a5bc-49ee-8a87-4f56e322139f.png">

Exposing something is easy, but once it's exposed, hiding it again would be a breaking change.

Reply:
Here's something concrete:

https://forum.rescript-lang.org/t/proposal-untagged-variants-for-rescript/4349/32?u=cristianoc

Just an observation about one specific part of json implementation in the compiler itself.

Generally speaking, what you gain is the ability to do pattern matching.
I think @zth has better nested samples, where by pattern matching you reach directly deep into the ison when required, as opposed to do a lot of wrapping an unwrapping with abstract accessors.

Reply:
To be even more concrete, do an exercise: validate json according to the following criteria:
1. it contains either a string, or a boolean, or an array
2. the string must be "true" or "false" or "True" or "False"
3. if an array, it must be an array of booleans

Write that with an abstract interface or with pattern matching. See if there's a difference.

Reply:
Once exposed, if we change the implementation of Json.t later, it will become a breaking change.

Reply:
The implementation of `Json.t` is json itself right? We're aiming for zero cost access.

Reply:
Opaque types are a programming concept where the internal structure of a type is hidden from the user, and access to the underlying data is only possible through a specific set of functions or methods. This is a form of data abstraction that can improve modularity, maintainability, and correctness in software design. However, opaque types come with their own advantages and disadvantages, which you should consider before deciding to use them:

Advantages:

1. Encapsulation: Opaque types hide the implementation details of a data structure, enforcing a clean separation between the interface and the implementation. This encapsulation helps promote modularity and makes code easier to understand, maintain, and refactor.

2. Controlled access: By exposing only a specific set of functions or methods to manipulate the underlying data, opaque types can help prevent unintended side effects and improve the overall robustness of the code.

3. Invariants enforcement: Opaque types can enforce invariants (conditions that must be true at all times) more effectively since they restrict direct access to the underlying data.

4. Easier API evolution: By using opaque types, you can change the underlying implementation without affecting client code, making it easier to evolve the API over time.

Disadvantages:

1. Performance overhead: Opaque types may introduce some performance overhead due to the additional function or method calls required for data manipulation. However, modern compilers can often optimize these calls, minimizing the performance impact.

2. Complexity: Introducing opaque types can make the code more complex, especially if the added abstraction is not necessary for the particular use case.

3. Indirect access: Opaque types may require additional copying or transformation of data when interacting with other parts of the system, which can be less efficient than direct access.

4. Learning curve: Developers who are unfamiliar with opaque types might face a learning curve when working with them, which could slow down development initially.

Should you always reach for one where possible?

It depends on the specific context and requirements of your project. Opaque types can be beneficial for promoting clean, modular, and maintainable code, but they might not be suitable for every situation. It's essential to consider the trade-offs and decide if the benefits of using opaque types outweigh the disadvantages in your particular use case. For example, you may choose to use opaque types when designing public APIs or when working on a large codebase that requires strong encapsulation and modularity. In contrast, opaque types might be less useful for small, performance-critical components where direct access to data is more important.

Reply:
Example of using an opaque type:

Suppose you're implementing a library for handling dates in a calendar system. You could create an opaque `Date` type that encapsulates the internal representation of a date (e.g., days since a specific epoch). Users of your library would interact with the `Date` type only through a set of provided functions, like `create_date`, `add_days`, `subtract_days`, etc., without ever knowing the underlying representation.

This approach provides several benefits:

1. You can change the internal representation of a date (e.g., switch from days since epoch to a combination of year, month, and day) without affecting client code.
2. You can enforce invariants like ensuring that the date is always valid (e.g., no February 30th).
3. You provide a cleaner, more maintainable interface to the users of your library.

Example of not using an opaque type:

Suppose you're implementing a small performance-critical module for manipulating 2D points in a graphics application. In this case, using a simple, non-opaque `Point` struct with public `x` and `y` coordinates might be more appropriate. This allows for direct access to the coordinates, which can result in better performance and less complexity.

In this scenario, the benefits of using an opaque type might not outweigh the disadvantages:

1. Performance is a critical concern, and introducing function or method calls for accessing the point's coordinates could lead to overhead.
2. The data structure is simple and has a low risk of exposing implementation details or violating invariants.
3. The codebase is small and focused, so the added abstraction of an opaque type might not provide significant benefits in maintainability or modularity.

Reply:
Trying to understand the concern about changing the implementation.
Still not getting it, but I guess there must be something.
Do you have concrete example of how the implementation of specifically json, will change overt time?

Reply:
( the perf angle is not just some theory: see how some tests, in this PR, have now become faster because some optimisation with options is unlocked)

Reply:
I didn't want to be a blocker for no reason. 🥲 
However, I still don't quite understand why exposing the original Js.Json.t type is an advantage. The ability to have zero-cost interop due to untagged unions and exposing the Js.Json.t type seem like separate topics. For example, using the Js.Json.string function to create a Js.Json.t of String(string) might be difficult for users to express, but it still seems that the value of Js.Json.t remains zero-cost interop to the value in runtime.

If we expose the definition of Js.Json.t, in fact, most of the interfaces of the Js.Json module become unnecessary. The Js.Json.t type value can be declared and used directly. If it is decided that this improves user experience and reduces the learning curve, I would like to suggest Js.Json2.t instead of the existing Js.Json.t.

Reply:
I will try to create a concrete example over the weekend. 🤔  I don't want to be a blocker for no reason, so let's continue the discussion afterward.

Reply:
I'll await your examples, this isn't time critical in that sense, and I'm curious to understand your points so we don't miss anything. However I honestly still don't get it. You're correct that most functions in `Js.Json` becomes unnecessary. But that's because support JSON can now be done first class in the language itself. 

We're unifying concepts here - instead of having to learn a bunch of quirks and special methods for interacting with something abstract, you can now replace that (if you want) with just using the language's two most powerful constructs; variants and exhaustive pattern matching. It's removing an entire middle layer that was more or less only created because it wasn't _possible_ before to do what we can do now.

But again, I'm interested in your examples. Don't want to be missing something here.

Reply:
For example, we have `Money` module and `Money.t` should not be a negative number. If `Money.t` is not defined as an opaque type, you can encounter Data integrity violation issues like the following:

It would be possible to create an instance of Money.t with a negative value, which is not allowed by the definition of the type.
```res
let negativeMoney: Money.t = -1
```
Similarly, using a non-opaque type for Js.Json.t can expose similar risks. If Js.Json.t is not opaque, you could create a Js.Json.t value without using the Js.Json.parse function, potentially defining and using a value that does not conform to the JSON specification.
```res
let notJsonString: Js.Json.String(string) = "The line break \n should be escaped in json"
```
By using an opaque type, you can prevent these issues and ensure that your data adheres to the intended constraints and specifications.

Reply:
I understand what you mean, but the current definition does confirm to the JSON spec, that's why we want to add it. So it is indeed impossible to construct a value using the proposed Js.Json.t that's not valid JSON. Provided it fully conforms to the spec, do you have other objections?

Reply:
How can we be sure? I mean, if we don't use the interface that forces us to create the Js.Json.t type value using the Js.Json.parse function which means the opaque type.

Reply:
Here is another example: circular reference object
```res
let circularReferencingObj: Js.Json.t = Js.Dict.(// making circular referencing object)->Js.Json.Object 
```

Reply:
While true, that type of problems are not new, not introduced by this change, and it's not something making the type abstract magically fixes unless we restrict it to _only_ allowing `JSON.parse` to return a `JSON.t`, which seems unpractical.

For example, there are a bunch of functions like `string: string => t` and `object_: Js.Dict.t<t> => t` in the JSON module right now that operates on the abstract type, but that has the same problem that you're describing, since they're just identity functions.

Reply:
I think we have exhausted the exploration here. Merging.

Reply:
While I respect the maintainer's decision and the fact that it's been merged, I still don't think the current situation justifies using non-opaque type for Js.Json.t. Rather, it might be possible to justify the existing functions due to the decision to use non-opaque types.

> there are a bunch of functions like string: string => t and object_: Js.Dict.t<t> => t in the JSON module right now that operate on the abstract type, but they have the same problem that you're describing, since they're just identity functions.

Reply:
Not sure what's the issue here.
What I see is a very talented contributor who has shipped awesome features executed extranet well: jsx4 and dynamic loading.
This issue does not seem very important.